### PR TITLE
Use ring multicore ccl ops in llama prefill

### PR DESCRIPTION
**Pull request by mvasiljevicTT**
### What's changed

- Using ring implementations for AG and RS in `llama_ccl.py` for Llama prefill, when RING_6U is on
- Making new semaphores and persistent buffers for those ops.
- Ring and line implementations can be mixed by using `LINE_AG` and `LINE_RS` env variables and setting `USE_LINE_AG` list in `llama_ccl.py`

### Example

To use line for all RS and ring for all AG except layer norm set RING_RS=1, USE_LINE_AG = {"LAYERNORM"}

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by mtairum** [comment]
Please run the TG pipelines 🙏 

Can you also add the 
`To use line for all RS and ring for all AG except layer norm set RING_RS=1, USE_LINE_AG = {"LAYERNORM"}` documentation to the README?

Or at least as a code comment.

File: models/demos/llama3_subdevices/tt/llama_ccl.py [hunk 1]
```diff
@@ -39,19 +61,33 @@ def __init__(
...
+                            [ttnn.create_global_semaphore(self.mesh_device, self.sub_device_crs, 0) for _ in range(12)]
+                        )
+
+                    else:
+                        self.from_semaphore_handles[i].append(
+                            ttnn.create_global_semaphore(self.mesh_device, self.sub_device_crs, 0)
+                        )
+                        self.to_semaphore_handles[i].append(
+                            ttnn.create_global_semaphore(self.mesh_device, self.sub_device_crs, 0)
+                        )
```
Comment (mtairum) [inline models/demos/llama3_subdevices/tt/llama_ccl.py]: For this part, can you add more clarity through comments on why were are creating 2 semaphores for AG, and 12 for RS when using ring, and just 1 `to` and 1 `from` when using line.

For future reference as well.

File: models/demos/llama3_subdevices/tt/llama_ccl.py [hunk 2]
```diff
@@ -413,6 +456,84 @@ def get_prefill_reduce_scatter_buffers(self):
             persistent_buffers_all[seqlen] = persistent_buffers
         return persistent_buffers_all
 
+    def padded_shape(self, output_shape, tile, num_devices, num_links, ag_input_dtype):
+        num_banks = 12
```
Comment (mtairum) [inline models/demos/llama3_subdevices/tt/llama_ccl.py]: Are these DRAM banks? If so, please add a comment that this is WH specific, as BH only has 8 banks.

File: models/demos/llama3_subdevices/tt/llama_ccl.py [hunk 3]
```diff
@@ -447,6 +568,57 @@ def get_prefill_all_gather_buffers(self):
...
+                )
+
+                tt_output_buffer = ttnn.as_tensor(
+                    torch.zeros(shape[1]),
+                    device=self.mesh_device,
+                    layout=ttnn.TILE_LAYOUT,
+                    dtype=ttnn.bfloat16 if key == "LAYERNORM" else ttnn.bfloat8_b,
+                    memory_config=ttnn.DRAM_MEMORY_CONFIG,
+                    mesh_mapper=ttnn.ReplicateTensorToMesh(self.mesh_device),
+                    cache_file_name=self.weight_cache_path / ("pb_ag_" + key + str(seqlen)),
```
Comment (mtairum) [inline models/demos/llama3_subdevices/tt/llama_ccl.py]: add a comment with `# output buffer is reused from line imlementation`

File: models/demos/llama3_subdevices/tt/llama_ccl.py [hunk 4]
```diff
@@ -447,6 +568,57 @@ def get_prefill_all_gather_buffers(self):
...
+                "LAYERNORM": [(1, 1, seqlen, 128), 4],
+            }
+            padded_buffers = {
+                key: (
+                    self.padded_shape(
+                        value[0], (32, 32), value[1], 4, ttnn.ttnn.bfloat16 if key == "LAYERNORM" else ttnn.bfloat8_b
+                    ),
+                    value[0],
+                )
+                for key, value in buffers_dict.items()
```
Comment (mtairum) [inline models/demos/llama3_subdevices/tt/llama_ccl.py]: can you add a comment on why is this padded?

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/demos/llama3_subdevices/tt/llama_ccl.py (hunk 2 above)
Comment (mvasiljevicTT) [reply models/demos/llama3_subdevices/tt/llama_ccl.py]: Padding removed completely

File: models/demos/llama3_subdevices/tt/llama_ccl.py (hunk 3 above)
Comment (mvasiljevicTT) [reply models/demos/llama3_subdevices/tt/llama_ccl.py]: Ring AG now uses all same buffers as line, so this function is removed

File: models/demos/llama3_subdevices/tt/llama_ccl.py (hunk 1 above)
Comment (mvasiljevicTT) [reply models/demos/llama3_subdevices/tt/llama_ccl.py]: added comment

**Review by mtairum — APPROVED** [review]
